### PR TITLE
Play monitor

### PR DESCRIPTION
**Pull request by Bvasilchikov**
### Description

This PR adds the initial support for a Play Monitor. 

### Review Checklist

**_It is the reviewers responsibility to also make sure every item here has been covered_**

- [ ] **Function & Class comments**: All function definitions (usually in the `.h` file) should have a javadoc style comment at the start of them. For examples, see the functions defined in `thunderbots/software/geom`. Similarly, all classes should have an associated Javadoc comment explaining the purpose of the class.
- [ ] **Remove all commented out code**
- [ ] **Remove extra print statements**: for example, those just used for testing
- [ ] **Resolve all TODO's**: All `TODO` (or similar) statements should either be completed or associated with a github issue

### CODE REVIEW

**Comment by itsarune** [comment]
can you target this PR to the `new_gameplay_staging_branch`

File: src/software/ai/hl/stp/play/monitor/play_monitor.cpp [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+        contains(world.field().enemyHalf(), world.ball().position()))
+    {
+        return -1.0;
+    }
+    if (attackingTeam == TeamPossession::ENEMY_TEAM &&
+        contains(world.field().enemyHalf(), world.ball().position()))
+    {
+        return -1.0;
+    }
+    LOG(WARNING) << "DribblerMode is invalid" << std::endl;
```
Comment (itsarune) [inline src/software/ai/hl/stp/play/monitor/play_monitor.cpp:75]: it's probably better to use a switch statement here

File: src/software/ai/hl/stp/play/monitor/play_monitor.h [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+    explicit PlayMonitor(const PlayIntent& initialIntent);
+
+    void startMonitoring(const World& initialWorld);
+    double endMonitoring(const World& finalWorld);
+    void updateWorld(const World& newWorld);
+    void updatePlayIntent(PlayIntent newIntent);
+
+   private:
+    double calculateIntentBallScore();
+    double calculateIntentActionScore();
```
Comment (itsarune) [inline src/software/ai/hl/stp/play/monitor/play_monitor.h:24]: these two functions aren't implemented unless I'm missing something

File: src/software/ai/hl/stp/play/monitor/play_monitor.cpp [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+        const auto ball_distance_travelled =
+            distance(finalWorld.ball().position(), world.ball().position());
+        // How far the ball has travelled along the x axis between updatedWorld and the
+        // old world, this indicates that generally the ball has move closer to the enemy
+        // goal. this could probably later be revised later by also considering the y
+        // value, when the ball is in the enemy third - at that point the play should try
+        // to funnel the ball closer to the goal
+        const auto ball_distance_travelled_along_field =
+            finalWorld.ball().position().x() - world.ball().position().x();
+        if (ball_distance_travelled > (4 * BALL_MAX_RADIUS_METERS))
```
Comment (itsarune) [inline src/software/ai/hl/stp/play/monitor/play_monitor.cpp:51]: `4*BALL_MAX_RADIUS_METERS` should be a constant

File: src/software/ai/hl/stp/play/monitor/play_monitor.cpp [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+    if (attackingTeam == TeamPossession::ENEMY_TEAM &&
+        contains(world.field().enemyHalf(), world.ball().position()))
+    {
+        return -1.0;
+    }
+    if (attackingTeam == TeamPossession::ENEMY_TEAM &&
+        contains(world.field().enemyHalf(), world.ball().position()))
+    {
+        return -1.0;
+    }
```
Comment (itsarune) [inline src/software/ai/hl/stp/play/monitor/play_monitor.cpp:74]: line 65 - 68 and line 70-73 logic is duplicated

**Review by itsarune — COMMENTED** [review]
the overall logic looks okay, but I left some comments about the implementation

**Comment by nimazareian** [comment]
Arun and I realized that this PR should actually be merged into the gameplay staging branch rather than master.

**Comment by Bvasilchikov** [comment]
closing as it was based off `master` instead of the `play_monitor_staging` branch. new progress is at #3169

**Comment by Bvasilchikov** [comment]
closing as it was based off `master` instead of the `new_gameplay_staging` branch. new progress is at #3169

**Pull request closed without merging**